### PR TITLE
Call flush_all twice when closing a client connection

### PR DESCRIPTION
**Pull request by catern**
The Merlin server computes responses in response to client requests.  Sometimes, it can
erroneously write a response from a previous client to a later client, causing multiple
responses to be returned.  I've figured out why this happens and fixed it; it's yet
another wacky collision between the Merlin server's somewhat unusual architecture and the
reality of Unix.

Specifically, the output channels might still have data in their buffers when we get to
the next client.  We call flush_all to try and prevent that, but that can fail with an IO
error, leaving the data in the buffer, and causing it to be written out for the next
client.

Ideally we'd just discard the buffered data for any channel that has an IO error, but I
doubt that's possible with the stdlib channels - I haven't looked in detail though.
Instead, we just flush the channels into the server's initial file descriptors.  That's
probably /dev/null or some logfiles, so it's an acceptable place to send that useless
data.

It's somewhat hard to reproduce this behavior, but repeatedly creating a client and then
killing it before sending all its input input is semi-reliable.

### CODE REVIEW

**Review by liam923 — APPROVED** [review]
lgtm

File: src/frontend/ocamlmerlin/ocamlmerlin_server.ml
Comment (ncik-roberts) [inline src/frontend/ocamlmerlin/ocamlmerlin_server.ml:26]: Having confusing data in log files might do more harm than what is claimed here. If I saw malformed-looking data in a log file, I would think something seriously wrong is going on and spend a lot of time investigating. Does the data actually end up in a log file? I wonder if we should log (and flush) some extra text to the fd, like "Finishing flushing stale data. In rare cases, this message will be preceded by some data that was intended to be sent to a client that recently crashed."

**Comment by ncik-roberts** [comment]
Overall this looks good to me and doesn't need explicit signoff before we merge. I left a minor comment about logging.

File: src/frontend/ocamlmerlin/ocamlmerlin_server.ml
Comment (liam923) [reply src/frontend/ocamlmerlin/ocamlmerlin_server.ml:26]: I think that's a good point and worth doing

Comment (catern) [reply src/frontend/ocamlmerlin/ocamlmerlin_server.ml:26]: BTW, stdout/stderr are /dev/null for ocamlmerlin-server on my box, which is probably the case everywhere, so it's unlikely to actually matter.

**Pull request merged**
